### PR TITLE
Rocksdb pageserver

### PR DESCRIPTION
**Pull request by knizhnik**

### CODE REVIEW

File: pageserver/src/lib.rs [hunk 1]
```diff
@@ -19,6 +21,8 @@ pub struct PageServerConf {
     pub daemonize: bool,
     pub interactive: bool,
     pub listen_addr: SocketAddr,
+	pub gc_horizon: u64,
+	pub gc_period: Duration,
```
Comment (ericseppanen) [inline pageserver/src/lib.rs:25]: Bad intent here. Please run `cargo fmt`.

File: pageserver/src/page_cache.rs [hunk 2]
```diff
@@ -226,51 +254,290 @@ pub struct CacheEntryContent {
...
+            rec: Bytes::from(dst),
+			main_data_offset
+        }
+    }
 }
 
 // Public interface functions
 
 impl PageCache {
+	fn do_gc(&self, conf: &PageServerConf) -> anyhow::Result<Bytes> {
```
Comment (ericseppanen) [inline pageserver/src/page_cache.rs:382]: This function creeps to the right so far, it's hard to follow the logic. Can we break it up into smaller functions?

File: pageserver/src/page_cache.rs [hunk 3]
```diff
@@ -226,51 +254,290 @@ pub struct CacheEntryContent {
...
+                shared.last_valid_lsn,
+                lsn
+            );
+        }
+
+        if shared.walreceiver_works {
+
+			while lsn > shared.last_valid_lsn {
+				// TODO: Wait for the WAL receiver to catch up
+				waited = true;
```
Comment (ericseppanen) [inline pageserver/src/page_cache.rs:506]: All the intents here are wrong. Please consider setting up your editor to automatically use the standard rustfmt style? There should be lots of guides online for how to do this.

File: pageserver/src/page_cache.rs [hunk 4]
```diff
@@ -8,29 +8,32 @@
...
-use std::collections::{BTreeMap, HashMap};
+use std::cmp::min;
+use std::collections::HashMap;
 use std::sync::atomic::AtomicU64;
 use std::sync::atomic::Ordering;
 use std::sync::{Arc, Condvar, Mutex};
 use std::thread;
 use std::time::Duration;
 use std::{convert::TryInto, ops::AddAssign};
+use rocksdb::*;
```
Comment (ericseppanen) [inline pageserver/src/page_cache.rs:26]: Can we please import only the symbols that are used?

File: pageserver/src/page_cache.rs [hunk 5]
```diff
@@ -613,80 +824,76 @@ impl PageCache {
...
-        let mut shared = self.shared.lock().unwrap();
-        let entry = shared.relsize_cache.entry(*rel).or_insert(0);
-        *entry
-    }
-
-    pub fn relsize_exist(&self, rel: &RelTag) -> bool {
-        let shared = self.shared.lock().unwrap();
-        let relsize_cache = &shared.relsize_cache;
-        relsize_cache.contains_key(rel)
+        return Ok(false);
```
Comment (ericseppanen) [inline pageserver/src/page_cache.rs:896]: `return` is not needed here.

File: pageserver/src/walredo.rs [hunk 6]
```diff
@@ -66,22 +69,23 @@ pub fn wal_redo_main(conf: &PageServerConf, timelineid: ZTimelineId) {
...
-        for _i in 1..100 {
+        for _i in 1..100000 {
             let request = walredo_channel_receiver.recv().unwrap();
 
             let result = handle_apply_request(&pcache, &process, &runtime, request);
             if result.is_err() {
                 // On error, kill the process.
+                error!("Kill wal redo process on error");
                 break;
             }
```
Comment (ericseppanen) [inline pageserver/src/walredo.rs:86]: I don't really understand the comment and message.

Is this saying "in the future we should kill the process"?  "Somebody already killed the process?"

File: pageserver/src/page_cache.rs [hunk 7]
```diff
@@ -226,51 +254,290 @@ pub struct CacheEntryContent {
...
+        buf.put_u32(self.dbnode);
+        buf.put_u32(self.relnode);
+        buf.put_u32(self.forknum as u32);
+    }
+    pub fn unpack(buf: &mut BytesMut) -> RelTag {
+        RelTag {
+            spcnode: buf.get_u32(),
+            dbnode: buf.get_u32(),
+            relnode: buf.get_u32(),
+            forknum: buf.get_u32() as u8,
```
Comment (ericseppanen) [inline pageserver/src/page_cache.rs:317]: When I see `as` used to truncate integers I usually ask whether there's a better way:
- use the smaller type in the first place
- use `try_into()` if we expect it to fit without truncation
If truncation is really what is needed, can you add a comment explaining why?

File: pageserver/src/page_cache.rs [hunk 8]
```diff
@@ -279,123 +546,46 @@ impl PageCache {
...
-                /* return Err("could not find page image")?; */
-            }
-            let (_key, entry) = entry_opt.unwrap();
-            entry_rc = entry.clone();
-
-            // Now that we have a reference to the cache entry, drop the lock on the map.
-            // It's important to do this before waiting on the condition variable below,
-            // and better to do it as soon as possible to maximize concurrency.
+        if entry_opt.is_none() {
+            static ZERO_PAGE: [u8; 8192] = [0 as u8; 8192];
```
Comment (ericseppanen) [inline pageserver/src/page_cache.rs:570]: `0 as u8` -> `0u8`

File: pageserver/Cargo.toml [hunk 9]
```diff
@@ -32,11 +32,13 @@ tokio-postgres = { git = "https://github.com/zenithdb/rust-postgres.git", rev="a
 postgres-types = { git = "https://github.com/zenithdb/rust-postgres.git", rev="a0d067b66447951d1276a53fb09886539c3fa094" }
 postgres-protocol = { git = "https://github.com/zenithdb/rust-postgres.git", rev="a0d067b66447951d1276a53fb09886539c3fa094" }
 postgres = { git = "https://github.com/zenithdb/rust-postgres.git", rev="a0d067b66447951d1276a53fb09886539c3fa094" }
+rocksdb = { git = "https://github.com/rust-rocksdb/rust-rocksdb.git" }
```
Comment (ericseppanen) [inline pageserver/Cargo.toml:35]: Why are we picking the upstream development branch? Do we need features that don't exist in any of the released versions?

If we need the bleeding-edge code, Please pin to a particular git rev (see the lines right above it). Otherwise, every time we build we may be getting a different version of the rocksdb code.

File: pageserver/src/page_cache.rs (hunk 4 above)
Comment (ericseppanen) [reply pageserver/src/page_cache.rs:26]: I think I changed my mind. The symbols needed from rocksdb include things named `Options`, `Direction`, and `IteratorMode`. Since those names don't communicate anything DB-related, I think the code would be more readable if we instead use `rocksdb::Options`, `rocksdb::Direction`, etc.

File: pageserver/src/walredo.rs (hunk 6 above)
Comment (knizhnik) [reply pageserver/src/walredo.rs:86]: Sorry, it is not my code. I think it is better to ask @hlinnaka  about his intention.

File: pageserver/src/page_cache.rs (hunk 2 above)
Comment (knizhnik) [reply pageserver/src/page_cache.rs]: Yeh, I also find this function too long.And it contains similar fragments of code which is better to extract in some functions. But unfortunately I didn't find nice way to do it, becasue of several mutable variables used and updated by this fragments: minbuf, maxbuf, minkey. maxkey. I do not want to move them to the object because them are used locally only in this thread. And I want to minimize buffer allocations. 

Let me think more about it. I just do not want to formally satisfy your requirement, for example by extracting loop body in separate functions.

File: pageserver/src/walredo.rs (hunk 6 above)
Comment (hlinnaka) [reply pageserver/src/walredo.rs:86]: I hope this clarifies it:
```
diff --git a/pageserver/src/walredo.rs b/pageserver/src/walredo.rs
index bf0159a..e46568f 100644
--- a/pageserver/src/walredo.rs
+++ b/pageserver/src/walredo.rs
@@ -76,11 +76,17 @@ pub fn wal_redo_main(conf: &PageServerConf, timelineid: ZTimelineId) {
 
             let result = handle_apply_request(&pcache, &process, &runtime, request);
             if result.is_err() {
-                // On error, kill the process.
+                // Something went wrong with handling the request. It's not clear
+                // if the request was faulty, and the next request would succeed
+                // again, or if the 'postgres' process went haywire. To be safe,
+                // kill the 'postgres' process so that we will start from a clean
+                // slate, with a new process, for the next request.
                 break;
             }
         }
 
+        // Time to kill the 'postgres' process. A new one will be launched on next
+        // iteration of the loop.
         info!("killing WAL redo postgres process");
         let _ = runtime.block_on(process.stdin.get_mut().shutdown());
         let mut child = process.child;
```

**Comment by hlinnaka** [comment]
The comment at the top of `page_cache.rs` is now out-of-date
